src/blueprints/tv2/action-factories/tv2-action-factory-provider.ts

```diff
@@ -0,0 +1,328 @@
+import {
+  Tv2GraphicsSplitScreenTimelineObjectFactory
+} from '../timeline-object-factories/interfaces/tv2-graphics-split-screen-timeline-object-factory'
+import {
+  Tv2RobotTimelineObjectFactory
+} from '../timeline-object-factories/interfaces/tv2-robot-timeline-object-factory'
+import {
+  Tv2TelemetricsTimelineObjectFactory
+} from '../timeline-object-factories/tv2-telemetrics-timeline-object-factory'
+
+
+interface ActionFactoryInstance<T> {
+  factory: T,
+  shouldFactoryBeRecreated: (configuration: Tv2BlueprintConfiguration) => boolean
+}
+
+export class Tv2ActionFactoryProvider {
+
+  private cameraActionFactoryInstance: ActionFactoryInstance<Tv2CameraActionFactory>
+  private remoteActionFactoryInstance: ActionFactoryInstance<Tv2RemoteActionFactory>
+  private transitionEffectActionFactoryInstance: ActionFactoryInstance<Tv2TransitionEffectActionFactory>
+  private audioActionFactoryInstance: ActionFactoryInstance<Tv2AudioActionFactory>
+  private graphicsActionFactoryInstance: ActionFactoryInstance<Tv2GraphicsActionFactory>
+  private videoClipActionFactoryInstance: ActionFactoryInstance<Tv2VideoClipActionFactory>
+  private videoMixerConfigurationActionFactoryInstance: ActionFactoryInstance<Tv2VideoMixerConfigurationActionFactory>
+  private splitScreenActionFactoryInstance: ActionFactoryInstance<Tv2SplitScreenActionFactory>
+  private replayActionFactoryInstance: ActionFactoryInstance<Tv2ReplayActionFactory>
+  private robotActionFactoryInstance: ActionFactoryInstance<Tv2RobotActionFactory>
+
+  public createCameraActionFactory(configuration: Tv2BlueprintConfiguration): Tv2CameraActionFactory {
+    this.cameraActionFactoryInstance = this.getUpdatedActionFactoryInstance(
+      this.cameraActionFactoryInstance,
+      () => {
+        return new Tv2CameraActionFactory(
+          this.createVideoMixerTimelineObjectFactory(configuration),
+          this.createAudioTimelineObjectFactory()
+        )
+      },
+      (c: Tv2BlueprintConfiguration): boolean => {
+        return this.didVideoMixerTypeChange(c, configuration)
+      },
+      configuration
+    )
+
+    return this.cameraActionFactoryInstance.factory
+  }
+
+  private getUpdatedActionFactoryInstance<T>(
+    actionFactoryInstance: ActionFactoryInstance<T> | undefined,
+    createFactoryCallback: () => T,
+    shouldFactoryBeRecreatedCallback: (configuration: Tv2BlueprintConfiguration) => boolean,
+    configuration: Tv2BlueprintConfiguration
+  ): ActionFactoryInstance<T> {
+    if (!!actionFactoryInstance && !actionFactoryInstance.shouldFactoryBeRecreated(configuration)) {
+      return actionFactoryInstance
+    }
+
+    actionFactoryInstance = {
+      factory: createFactoryCallback(),
+      shouldFactoryBeRecreated: shouldFactoryBeRecreatedCallback
+    }
+
+    return actionFactoryInstance
+  }
+
+  private didVideoMixerTypeChange(oldConfiguration: Tv2BlueprintConfiguration, newConfiguration: Tv2BlueprintConfiguration): boolean {
```

Comment:
Don't we usually do it from a past perspective and say `hasVideoMixerTypeChanged` in these cases? In any case i actually don't dislike this, we usually just don't see it, i just want to make sure.

Reply:
`did` is also past tense, but you are right, we normally have been using `has`.
I think both works, so if there is a desire to change it to `has`, I don't mind :)